hashes/src/macros.rs

```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Public macros.
+
+/// Adds trait impls to a bytelike type.
+///
+/// Implements:
+///
+/// * `str::FromStr`
+/// * `fmt::{LowerHex, UpperHex}` using `hex-conservative`.
+/// * `fmt::{Display, Debug}` by calling `LowerHex`
+/// * `serde::{Deserialize, Serialize}`
+/// * `AsRef[u8; $len]`
+/// * `AsRef[u8]`
+/// * `Borrow<[u8]>`
+/// * `slice::SliceIndex<[u8]>`
+///
+/// Arguments:
+///
+/// * `ty` - The bytelike type to implement the traits on.
+/// * `$len` - The number of bytes this type has.
+/// * `$reverse` - `true` if the type should be displayed backwards, `false` otherwise.
+/// * `$gen: $gent` - generic type(s) and trait bound(s).
+#[macro_export]
+macro_rules! impl_bytelike_traits {
```

Comment:
If the intention is to use this in `primitives` then we need to figure out parse errors in `FromStr` impl.

But also, it looks like the macro could be in `internals`.